File: kotlin-node/karakum/src/jsMain/resources/nameResolvers/resolveInterfacePropertyPropertyItemPropertyItemName.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default (node, context) => {
    const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())
    const getParent = typeScriptService?.getParent.bind(typeScriptService) ?? (node => node.parent)

    const subTypeReference = getParent(node)
    if (!subTypeReference) return null
    if (!ts.isTypeReferenceNode(subTypeReference)) return null
    if (!ts.isIdentifier(subTypeReference.typeName)) return null
    if (subTypeReference.typeName.text !== "Array") return null

    const subSubProperty = getParent(subTypeReference)
    if (!subSubProperty) return null
    if (!ts.isPropertySignature(subSubProperty)) return null
    if (!ts.isIdentifier(subSubProperty.name)) return null

    const subSubPropertyName = subSubProperty.name.text

    const subTypeLiteral = getParent(subSubProperty)
    if (!subTypeLiteral) return null
    if (!ts.isTypeLiteralNode(subTypeLiteral)) return null

    const typeReference = getParent(subTypeLiteral)
    if (!typeReference) return null
    if (!ts.isTypeReferenceNode(typeReference)) return null
    if (!ts.isIdentifier(typeReference.typeName)) return null
    if (typeReference.typeName.text !== "Array") return null

    const subProperty = getParent(typeReference)
    if (!subProperty) return null
    if (!ts.isPropertySignature(subProperty)) return null
    if (!ts.isIdentifier(subProperty.name)) return null

    const subPropertyName = subProperty.name.text

    const typeLiteral = getParent(subProperty)
    if (!typeLiteral) return null
    if (!ts.isTypeLiteralNode(typeLiteral)) return null

    const property = getParent(typeLiteral)
    if (!property) return null
    if (!ts.isPropertySignature(property)) return null
    if (!ts.isIdentifier(property.name)) return null

    const propertyName = property.name.text

    const interfaceNode = getParent(property)
    if (!interfaceNode) return null
    if (!ts.isInterfaceDeclaration(interfaceNode)) return null

    const parentName = interfaceNode.name.text

    return `${karakum.capitalize(parentName)}${karakum.capitalize(propertyName)}${karakum.capitalize(subPropertyName)}Item${karakum.capitalize(subSubPropertyName)}Item`
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/nameResolvers/resolveInterfacePropertyPropertyItemPropertyItemName.kt
package node.karakum.nameResolvers

import io.github.sgrishchenko.karakum.extension.Context
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import node.karakum.util.nullable
import typescript.Node
import typescript.isIdentifier
import typescript.isInterfaceDeclaration
import typescript.isPropertySignature
import typescript.isTypeLiteralNode
import typescript.isTypeReferenceNode

fun resolveInterfacePropertyPropertyItemPropertyItemName(node: Node, context: Context) = nullable {
    val typeScriptService = ensureNotNull(context.lookupService<TypeScriptService>(typeScriptServiceKey))

    val subTypeReference = ensureNotNull(typeScriptService.getParent(node))
    ensure(isTypeReferenceNode(subTypeReference))

    val subTypeName = subTypeReference.typeName
    ensure(isIdentifier(subTypeName))
    ensure(subTypeName.text == "Array")

    val subSubProperty = ensureNotNull(typeScriptService.getParent(subTypeReference))
    ensure(isPropertySignature(subSubProperty))

    val subSubPropertyNameNode = subSubProperty.name
    ensure(isIdentifier(subSubPropertyNameNode))
    val subSubPropertyName = subSubPropertyNameNode.text

    val subTypeLiteral = ensureNotNull(typeScriptService.getParent(subSubProperty))
    ensure(isTypeLiteralNode(subTypeLiteral))

    val typeReference = ensureNotNull(typeScriptService.getParent(subTypeLiteral))
    ensure(isTypeReferenceNode(typeReference))

    val typeName = typeReference.typeName
    ensure(isIdentifier(typeName))
    ensure(typeName.text == "Array")

    val subProperty = ensureNotNull(typeScriptService.getParent(typeReference))
    ensure(isPropertySignature(subProperty))

    val subPropertyNameNode = subProperty.name
    ensure(isIdentifier(subPropertyNameNode))
    val subPropertyName = subPropertyNameNode.text

    val typeLiteral = ensureNotNull(typeScriptService.getParent(subProperty))
    ensure(isTypeLiteralNode(typeLiteral))

    val property = ensureNotNull(typeScriptService.getParent(typeLiteral))
    ensure(isPropertySignature(property))

    val propertyNameNode = property.name
    ensure(isIdentifier(propertyNameNode))
    val propertyName = propertyNameNode.text

    val interfaceNode = ensureNotNull(typeScriptService.getParent(property))
    ensure(isInterfaceDeclaration(interfaceNode))
    val parentName = interfaceNode.name.text

    "${parentName.replaceFirstChar { it.titlecase() }}${propertyName.replaceFirstChar { it.titlecase() }}${subPropertyName.replaceFirstChar { it.titlecase() }}Item${subSubPropertyName.replaceFirstChar { it.titlecase() }}Item"
}
